src/expr/impl/src/scalar/case.rs

```diff
@@ -106,6 +107,116 @@ impl Expression for CaseExpression {
     }
 }
 
+/// With large scale of simple form match arms in case-when expression,
+/// we could optimize the `CaseExpression` to `ConstantLookupExpression`,
+/// which could significantly facilitate the evaluation of case-when.
+#[derive(Debug)]
+struct ConstantLookupExpression {
+    return_type: DataType,
+    arms: HashMap<ScalarImpl, BoxedExpression>,
+    fallback: Option<BoxedExpression>,
+}
+
+impl ConstantLookupExpression {
+    fn new(
+        return_type: DataType,
+        arms: HashMap<ScalarImpl, BoxedExpression>,
+        fallback: Option<BoxedExpression>,
+    ) -> Self {
+        Self {
+            return_type,
+            arms,
+            fallback,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Expression for ConstantLookupExpression {
+    fn return_type(&self) -> DataType {
+        self.return_type.clone()
+    }
+
+    async fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let input_len = input.capacity();
+        let mut builder = self.return_type().create_array_builder(input_len);
+
+        // TODO: Ensure if the last column matches the actual result generated from the operand
+        // Get the last column
+        let column_len = input.columns().len();
+        let column = input.column_at(column_len - 1);
+
+        for i in 0..input_len {
```

Comment:
And note that, since we **inevitably** need to traverse through the input **one by one**, rather than vectorizing (`CaseExpression` originally perform this), the potential optimized scenario is that when the number of arms in the defined case-when expression are significantly large and (possibly) the input data does not contain too much rows.

Reply:
I think you can use a sql to benchmark it. For example, `select  a giant case when ... from generate_series(1, 100000) `;

Reply:
Indeed! I will first try manually benchmarking such case by the time recorded by risedev.